src/lib.rs

```diff
@@ -0,0 +1,155 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::cell::UnsafeCell;
+
+pub mod impls;
+
+pub trait ConstInit {
+    /// Create a new instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+}
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a ScopedRawMutex and contains the protected data.
+///
+/// # Safety
+///
+/// ScopedRawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
+pub unsafe trait ScopedRawMutex {
```

Comment:
Do you think it's worth including an `is_locked` method as well? I did in `maitake-sync`, and it's required if any mutex wrapping a `ScopedRawMutex` wants to provide such a method. I can certainly live without it though, I don't actually think it's terribly useful IRL...